### DISCUSSION TITLE
30B model now needs only 5.8GB of RAM? How?

### DISCUSSION THREAD

**Discussion by pugzly** [ggml-org/llama.cpp#638] [General]
(Edit: apologies, I should have clarified initially I'm running on Linux OS. I didn't realize it might not be obvious from the screenshot alone for a non-Linux users.All tests are done on Ubuntu based Linux Mint 21.1)

I've been only playing with 30B model so far, since neither 7B nor 13B were very engaging.
As recently as yesterday 30B model fill just close to 30GB, but today's release now runs fine with less than 6GB (and that's with system memory usage).

Initially I thought it must be a bug, but I couldn't notice any quality loss in responses, and then I saw there was some major change introduced only hours ago, but the fundamentals of those changes are a little over my head.

Maybe someone smarter than me can at least roughly explain, in basic terms (if that's even possible at all), how memory usage dropped  5 times overnight?

Thanks a lot in advance.

![Screenshot_from_2023-03-30_22-57-53](https://user-images.githubusercontent.com/56161932/228975945-5506174f-8d16-4e23-8015-d1d1ac3032df.jpg)

**Comment by Green-Sky** [comment]
it did actually not, you are not seeing the actual ram usage, because the os now counts it as filesystem cache. Since https://github.com/ggerganov/llama.cpp/pull/613 the model is a memory mapped file. It does not do the impossible, it's just that gnome-system-monitor does not show cached files. :smile: 

Try `htop` or similar.

![image](https://user-images.githubusercontent.com/2938071/228984466-cbb04634-9461-489e-a509-8601f4fd13c3.png)

**Reply by pugzly** [reply]
Ah that makes sense. It didn't cross my mind to verify with htop that voodoo I was seeing last night.
But it seems I'll have to use that to monitor mem usage from now on.

Thanks a lot!

**Reply by jart** [reply]
However the memory usage of the process is still much lower. The operating system isn't lying to you. What that means is you can now have more LLaMA processes running at the same time. It means you can load much bigger models. My best guess is that my recent change caused "true" memory usage to decrease by 2x, because earlier, you had to make a full copy. Now you don't need to copy at all.

**Reply by jart** [reply]
Another thing you can do, is pass the `--mlock` flag which will lock the file pages into memory much harder, and therefore should give you the thing you were expecting. Please note it shouldn't matter, unless your system is reacting to memory pressure. Results may vary.

**Reply by jart** [reply]
Actually I take back what I said. I have better information now that I took a closer look at 30B specifically. I'll write a new response. @Green-Sky is mistaken.

**Reply by nonetrix** [reply]
For me on NixOS it seems htop doesn't show the real memory as well, however it does show it in the process list. Here is me running a 70B model with 4 bits, is there a way to make it count against the main counter and in btop as well ideally? I really don't like this behavior to begin with, I thought I could trust the memory usage bar to be accurate why is this default?!
![image](https://github.com/ggerganov/llama.cpp/assets/45698918/8e504dae-e4d7-40f6-8a6d-ef6dd306d437)

**Comment by jart** [comment]
Hi. I'm the author of #613 which is what made this improvement. I'm glad you're happy with the fact that LLaMA 30B (a 20gb file) can be evaluated with only 4gb of memory usage!

The thing that makes this possible is that we're now using mmap() to load models. This lets us load the read-only weights into memory without having to read() them or even copy them. The operating only has to create page table entries which reserve 20GB of virtual memory addresses. Crudely speaking, mapping 20GB of RAM requires only 40MB of page tables (`(20*(1024*1024*1024)/4096*8) / (1024*1024)`). The individual pages aren't actually loaded into the resident set size on Unix systems until they're needed.

Here's why lazy loading of memory matters. LLaMA 30B appears to be a *sparse* model. While there's 20GB of weights, depending on your prompt I suppose only a small portion of that needs to be used at evaluation time. It should be possible to measure exactly how many lazy loads are happening using a tool I wrote called rusage.com which I wrote a blog post about two weeks ago: https://justine.lol/rusage/

If I run 30B on my Intel machine:

```
$ rusage doas ./main -m ./models/30B/ggml-model-q4_0.bin -c 512 -t 12 -n 40 -s 1680277445 -p '## Question: What is best in life? ## Jeeves: '
...
llama_print_timings:        load time =  4067.39 ms
llama_print_timings:      sample time =    18.92 ms /    40 runs   (    0.47 ms per run)
llama_print_timings: prompt eval time =  5407.50 ms /    16 tokens (  337.97 ms per token)
llama_print_timings:        eval time = 25576.09 ms /    39 runs   (  655.80 ms per run)
llama_print_timings:       total time = 32382.17 ms
RL: took 33,484,182µs wall time
RL: ballooned to 23,024,644kb in size
RL: needed 373,356,606µs cpu (0% kernel)
RL: caused 411,522 page faults (100% memcpy)
RL: 793 context switches (24% consensual)
```

As we can see, 400k page faults happen, which means only 1.6 gigabytes (`(411522 * 4096) / (1024 * 1024)`) of the 20 gigabyte weights file actually needed to be used.

Now, since my change is so new, it's possible my theory is wrong and this is just a bug. I don't actually understand the inner workings of LLaMA 30B well enough to know *why* it's sparse. Maybe we made some kind of rare mistake where llama.cpp is somehow evaluating 30B as though it were the 7B model. Anything's possible, however I don't think it's likely. I was pretty careful in writing this change, to compare the deterministic output of the LLaMA model, before and after the Git commit occurred. I haven't however actually found the time to reconcile the output of LLaMA C++ with something like PyTorch. It'd be great if someone could help with that, and possibly help us know why, from more a data science (rather than systems engineering perspective) why 30B is sparse.

Until then, I hope you enjoy it! You can now run a bigger, badder, better model on your PC without compromising other programs. The recent change also means you can run multiple LLaMA `./main` processes at the same time, and they'll all share the same memory resources. Similar to how many executables are able to mmap() the same shared object dependencies into memory.

**Reply by pugzly** [reply]
Thanks a lot for your work and for the in depth explanation.
I haven't had time last night to fully test and check it all, didn't try to load 60B model (I might try tmrw tho).
But I definitely tested responses given by the loaded model, so based on that it was clear the 30B model was responding, not a 7B or 13B ones.
So it seems there is no issue with the model, I'll just have to test now how is my system managing memory with this update, if there is any quirks I should be aware of. Probably over the weekend when I have a bit more free time.

**Reply by Green-Sky** [reply]
this is actually a nice way of being mistaken :D

**Reply by j-f1** [reply]
Interesting! I have also noticed some slowdowns (with 13b) that I attribute to the model having to load up more weights mid-response. Would be interesting to have some sort of heat map that shows which parts of the model each token touched!

**Reply by jart** [reply]
There's actually a system call we can use to build a heat map called mincore(). Someone should sample the process at like 10ms intervals and create a video visualization of the pages that are in core.

**Reply by blackrabbit17** [reply]
@jart this is really amazing work, 

Have you given much thought to monitoring the runtime weight usage in a rigorous way to influence weight pruning? - Just curious for your thoughts on this

**Reply by blackhole89** [reply]
@jart Is it possible that transparent hugepages were involved (so the assumption that one page fault = 4KB does not hold)? I remember previously playing around with the allocations and finding that on my (recent-ish Linux kernel) system the big mallocs automatically used them already.

**Reply by zygi** [reply]
I was intrigued by this because the 30B model definitely isn't sparse in any way. So my intuition is, either we're monitoring mmap faults wrong (it's famously difficult to do it right), or there's a bug in inference code.

Specifically, one complication is that the kernel normally does readahead by default when encountering a page fault (https://unix.stackexchange.com/questions/533811/weird-major-page-fault-number-when-reading-sequentially-randomly-in-mmap-regio). However, even after providing the right hint to mmap to disable readahead, I'm still getting unexpectedly low numbers.

(also I've confirmed that the page size is the default 4K)

**Reply by rlrs** [reply]
This is very intriguing, but unless I am reading this discussion wrong, there is a claim that not all the weights are being read *eventually* during a forward pass? If so, that must be a bug in either the inference code or in the method of measuring the amount of data loaded.

**Reply by atondwal** [reply]
@zygi Is what you did comparing `POSIX_FADV_RANDOM` and/or `POSIX_FADV_DONTNEED` with `POSIX_FADV_SEQUENTIAL` and/or `POSIX_FADV_NORMAL`? (maybe try with THP set to `never`?)

It's definitely strange that all the weights aren't being read in. Even if this was a "sparse" model in the ML sense, the weights would be stored in a "sparse matrix" representation in the file, which would amount to a dense read.

**Reply by zygi** [reply]
> there is a claim that not all the weights are being read eventually during a forward pass?

yup, from the comment above: 
> As we can see, 400k page faults happen, which means only 1.6 gigabytes ((411522 * 4096) / (1024 * 1024)) of the 20 gigabyte weights file actually needed to be used.

but I'm pretty sure that's wrong -- `top` actually reports >20gb being used by the `main` binary, as expected, and running it in a 10gb virtual machine slows inference down massively (since now weights have to be re-read from memory for every token)

**Reply by zygi** [reply]
(though no matter how hard I try, linux keeps telling me an unreasonably low number of page faults happened. I think the reason is aggressive readahead which even `POSIX_MADV_RANDOM` doesn't disable. in any case I don't think the reported effect is real.)

**Reply by CoderRC** [reply]
On windows from https://github.com/ggerganov/llama.cpp/issues/103#issuecomment-1487784446 I see how much memory it uses for me 2 gb from 26 gb fp32.

**Reply by Doublespaced1** [reply]
So... is this something we could apply to models that run on GPUs as well? It would be extremely nice to have high-speed text generation from a 30b or even 65b model on more generally available consumer hardware.

**Reply by apage43** [reply]
> So... is this something we could apply to models that run on GPUs as well? It would be extremely nice to have high-speed text generation from a 30b or even 65b model on more generally available consumer hardware.

the advantage here comes mostly from avoiding unnecessary copying - with the change the memory the data is initially loaded into from disk is the exact memory that is used for inference, if you are running a model larger than the memory you have you will still wind up unloading and reloading it for each inference pass

the closest thing for GPU is the ability to directly copy data from NVMe drives to GPUs without having to go through system RAM on the way there - MS's DeepSpeed implements "NVMe offload" which can allow you to use that to run models larger than your available VRAM by swapping in only the layers being currently used, but its mostly used for training where batching means the throughput stays decent even if the actual time to evaluate the model once takes a lot longer, it likely wouldn't be worthwhile for single user inference. as you'll still need to load every layer at least once for every token.

**Reply by diimdeep** [reply]
Current SSD's achieve 7.5 GB/s+ read speed, opposed to older SSD from 2013 with 500 MB/s, so performance will drastically differ depending on your system specs. 
Also, try to run `vmmap <pid>` on process, to see more detailed memory and swapping stats.

**Reply by InconsolableCellist** [reply]
> Hi. I'm the author of #613 which is what made this improvement.

Credit where it's due, co-author

**Reply by acatovic** [reply]
@jart wow this is amazing work Justine!

You asked about the sparsity characteristics, from the model perspective. I think Tim Dettmers explains this very well here: https://timdettmers.com/2022/08/17/llm-int8-and-emergent-features. It is especially pronounced once you go beyond 7B parameter attention-based models, and probably why you are seeing this behaviour with 30B model here. If I understood it correctly:

* One characteristic of attention-based models, even smaller ones (with as few as 100M parameters), is the emergence of "outlier features", which are probably responsible for giving the models that incredible expressive ability and a knack for nuance
* Once you go above 7B parameters, a "phase shift" occurs, where these outlier features become even greater in number, and present across all transformer layers
* However, they start to coordinate through only a small number of hidden dimensions
* Feed-forward layers become highly dense, HOWEVER, the attention layers become extremely sparse, almost binary in nature

I think this is the same reason why certain low precision quantization like 4-bit GPTQ work extremely well on very large models!

**Reply by Tolga-Karahan** [reply]
Could someone explain this calculation: `(20*(1024*1024*1024)/4096*8) / (1024*1024)`. My OS is a bit rusty. :smile:

**Reply by chowder** [reply]
> Could someone explain this calculation: `(20*(1024*1024*1024)/4096*8) / (1024*1024)`. My OS is a bit rusty. 😄

`(20GB / 4KB Page size * 8 bytes per PTE) / 1MB` = 40MB

**Reply by khengari77** [reply]
Interesting, is it possible to port this to whisper.cpp?

**Comment by CoderRC** [comment]
Wow, mmap has a side affect of not loading the file data until it is used even I am surprised why this project loaded the whole model not what it needed. This looks like a hint that rearranging the model in memory might potentially give great performance improvements.

**Reply by danuker** [reply]
In hindsight, it's obvious. But it takes experience to get *insight*.

**Comment by raphael-sch** [comment]
Pretty sure it’s the input embedding weights matrix that benefits a lot from lazy loading. It contributes a significant amount to the model’s size. Each of the 32k tokens in the vocabulary has its row in the matrix but only a tiny subset of tokens are needed for each query. 

BTW, I was wrong. While there is sparsity in the embeddings, they are not a significant amount of the model's weights

**Reply by scottjmaddox** [reply]
That's true for the input, but not at the output, assuming the embedding layer is shared. At the output, getting the logits requires a dense matmul.

**Reply by eqy** [reply]
Yes, the output layer https://github.com/facebookresearch/llama/blob/57b0eb62de0636e75af471e49e2f1862d908d9d8/llama/model.py#L215
looks like a dim * vocab size chonker of a layer, which I would assume would needed to be loaded densely. However, this alone would only represent `6656 * 32000` parameters for the 33B model, which seems like it wouldn't be on the same order magnitude as the savings reported here.

EDIT: the input embedding size looks to be similar so I'm not sure that would explain such a dramatic savings:
https://github.com/facebookresearch/llama/blob/57b0eb62de0636e75af471e49e2f1862d908d9d8/llama/model.py#L206

**Reply by josemlopez** [reply]
The embeddings are taking only a marginal portion of the total size for a model of 30B; the lion's share belongs to the FFN (which makes sense).

Although is not the same architecture, this will give some idea of the order of magnitude: https://www.lesswrong.com/posts/3duR8CrvcHywrnhLo/how-does-gpt-3-spend-its-175b-parameters,
I agree that the embedding layer is a good hypothesis, but it is not enough.

**Reply by raphael-sch** [reply]
agree that it doesn’t explain the huge drop in memory footprint reported here. Also likely that the embedding matrix is loaded as a whole but accessed sparsely.

**Reply by apage43** [reply]
The input embedding layer would be sparsely accessed as that's just how embedding layers usually work, but the rest of the model is definitely fully dense - what's actually happened is just that MAP_SHARED just makes memory accounting confusing - since multiple processes with shared mappings of the same file are using the same memory, its not counted against any single process - it will be included in system wide counters as "cached", and you can use `mincore` as I showed below to check how much of any given file is cached. If you try to run a model larger than your RAM the file will not be able to be entirely cached and there will be persistent disk activity during inference as each layer will be re-read on use.

**Reply by josemlopez** [reply]
Thanks for the clarification, I just saw your graph. ~~Does it mean we have virtually gain nothing?~~

**Reply by apage43** [reply]
not quite - the models do load *faster* now, you can run multiple instances without loading the model multiple times, and you don't need to wait for it to be read from disk on subsequent runs as caching it is now managed by the OS, and although its slow, running models larger than RAM just reads your disk a lot, but does not need additional swap space and therefore should not need to *write* to your disk

**Reply by josemlopez** [reply]
Sure. I will rephrase. I meant the model is not using only the 6GB it seemed.

**Comment by leuc** [comment]
It would be great, to see the same memory use improvement from `mmap()` with whisper.cpp

**Comment by apage43** [comment]
I don't think there's any actual sparseness happening - I did a quick check with `mincore()` as suggested - quick and dirty `mincore` sampler here https://gist.github.com/apage43/ef502602a511ccff73a8882677e54627 - and plotted the # of resident pages over time, generation only starts when the model is fully paged in and it remains fully paged in:

![59039298-80ea-4c8f-b786-0c53b946803e](https://user-images.githubusercontent.com/169252/229258182-5901c45f-ef4f-4a9c-bca6-9532a9a0c6b0.png)

**Reply by jart** [reply]
Hmm. OK in that case what's probably happening is the tools like htop are just subtracting shared memory from resident memory when reporting it to the user.

**Reply by cztomsik** [reply]
maybe for the (input) embeddings part but that's it, everything else is always used for computation, it's either reporting or something else.

**Comment by qdwang** [comment]
In my test of using 30B gptq 4bit model under windows with 16g memory, it still consumes more than 16gb...And during the very slow processing, the ssd reading speed is at 100% capacity.

**Reply by dim-ask** [reply]
Same in my 16GB m1 macbook air. It seems it rereads the whole weights file from the disk again to compute each new token (as it prints every ~20GB disk read), no swap memory written, it is a pity. The 13B quantised model runs faster than before, and subsequent calls take very little time to load as the file is already cached in RAM.

**Reply by LoganDark** [reply]
Windows doesn't have mmap and I don't know what macOS does (macOS does have mmap), but this post is about Linux.

**Reply by dim-ask** [reply]
1. There is no specific mention about linux in the OP
2. Others here report similar behaviour on linux after surpassing the confusion caused by reading the monitoring tools (as one would expect, unless you have something different to report) - but I am happy to also check it on linux later.

**Reply by jart** [reply]
Windows reports memory usage different than Linux tools do. You should see on Windows using Process Explorer the "physical memory" should increase by the full amount, however the "physical commit" won't increase much. That's probably the most analogous comparison to what Linux is doing.

**Reply by jart** [reply]
Windows does have mmap() it just goes by a different name. See our polyfill code here:

https://github.com/ggerganov/llama.cpp/blob/a717cba8440b380f43cd3e2510862fc1ea3de9a2/llama.cpp#L310-L340

**Comment by rabidcopy** [comment]
I'm becoming somewhat doubtful of how much `mmap` actually reduces RAM usage. Despite what the OS reports, it's clear in my case on a RAM-limited system, the 30B model will go into swap and result in extremely slow speeds. The swap very clearly filling up despite free RAM being shown as available. I think to say `mmap` reduces memory usage is a slight misnomer. Feel free to be prove me wrong and say my experience doesn't reflect reality though. While there may be some technicalities on the language surrounding it, I'm leaning towards what @Green-Sky states and that it's not magic. It's not a bad thing to have and does make certain use cases much quicker and I'm not trying to be ruthlessly abrasive, but by no means is it enabling lower end systems to load models they previously were unable to. At least not in a usable state. I don't think waiting minutes per token is realistic.

**Reply by trollkotze** [reply]
Yeah, l can confirm, looks like that's what's happening for me, too. While previously all the 7 cores I assigned to llama.cpp were busy with 100% usage and almost all of my 30GB actual RAM used by it, now the cpu cores are only doing very little work, mostly waiting for all the loaded data in swap, apparently. Really weird.
I found that I can make it use real RAM again by starting llama.cpp with the parameter "--mlock", using "locked memory", and its performance seems to be as it was before then. However, I am only able to do that as root, since there seem to be fixed limits in place how much "locked memory" a normal user can use.
I have no clue about this concept of "locked memory". It's quite puzzling that the earlier version just used up all my RAM, refusing to use any swap at all (memory usage of llama.cpp shown as "pinned memory", i.e. non-swappable in gnome-system-monitor) when I ran it as a normal user, but now I need extra privileges to explicitly ask for "locked memory" to use.

**Reply by comex** [reply]
> I'm becoming somewhat doubtful of how much `mmap` actually reduces RAM usage. 

In reality [edit: almost] the entire model is needed for inference, so mmap doesn’t reduce RAM usage at all.  It’s purely a measurement artifact.  The exception is while loading the model at startup.  Reading without mmap can temporarily put more memory pressure on the rest of the system because chunks of the file will temporarily exist in memory twice (once in the kernel’s page cache, once in the application’s buffer).  Not sure how significant this effect is; it’s separate from the main benefit of mmap, which is the ability to reuse loaded data across multiple runs.

EDIT: I lied; @raphael-sch is correct that the input embeddings matrix is accessed sparsely.  But that is only 250-500MB, depending on the model size.

**Reply by comex** [reply]
> I have no clue about this concept of "locked memory". It's quite puzzling that the earlier version just used up all my RAM, refusing to use any swap at all (memory usage of llama.cpp shown as "pinned memory", i.e. non-swappable in gnome-system-monitor) when I ran it as a normal user, but now I need extra privileges to explicitly ask for "locked memory" to use.

By default, Linux treats mmapped pages as somewhat lower priority to keep in memory than anonymous (i.e. non-mmapped) pages. This is because anonymous pages have to be actually written to swap, whereas with mmapped pages (at least if they’re unmodified, as is the case here), the kernel can just toss them away, and read them back from their existing location in disk when they’re needed again.  This page has more detail:

https://biriukov.dev/docs/page-cache/4-page-cache-eviction-and-page-reclaim/

It’s a simple heuristic that in this case is counterproductive.  You could try adjusting `vm.swappiness`, but it’s probably better to just use mlock.

**Reply by FNsi** [reply]
I think so, my ram not enough for loading q4_1 65b, in previous version it just crash but after mmp, it can running, and I checked my swap be taken like 16g. At least it work!

**Comment by chadbrewbaker** [comment]
To make this scale on clusters of low memory hardware you  need either:
* [HTTP range queries](https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests) like AWS S3. This supports workflows complex as [walking sqlite b-trees](https://github.com/dacort/athena-sqlite/tree/master/lambda-function)
* Remote Direct Memory Access (RDMA). Message Passing Interface (MPI) or memcached.

1. What partition of the model makes sense?

2. Could inference be refactored to [parallel prefix](http://courses.csail.mit.edu/18.337/2004/book/Lecture_03-Parallel_Prefix.pdf)?

If so you could evaluate the model in parallel for speedup. Same loop in PicoGPT for readability.
 
https://github.com/ggerganov/llama.cpp/blob/d8d4e865cd481b18f10508ffee35db903767ef5c/examples/main/main.cpp#L243

https://github.com/jaymody/picoGPT/blob/3b7f4d180bb125a76b2f4f7b3a74268e5ec5f131/gpt2_pico.py#L45

Example of [arbitrary length Roman Numeral evaluation](https://gist.github.com/chadbrewbaker/4542999) as prefix sum. Hardest part is usually coming up with an identity state.

--update--
ChatGPT3.5 suggestion for refactor of evaluation to prefix sum over tokens. Looks somewhat sane?

https://gist.github.com/chadbrewbaker/ffe95290fc945af63611693688dfe54d

**Comment by 4uniquemind** [comment]
Please, do it with Koala LLM!

**Comment by dustyny** [comment]
Does this work with the [4 bit versions?](https://github.com/qwopqwop200/GPTQ-for-LLaMa). Would there be any advantage to running the 4 Bit 65B model on a consumer card (I have 24GB).

**Reply by trollkotze** [reply]
@dustyny
The question in the title can be answered with "no". It was a misunderstanding.
Llama.cpp does not run on GPU, so your graphics card won't help you. The relevant metric is your normal system RAM.
4 Bit 65B runs fine with 64GB of RAM. I even got it running on 32GB with zram-swap configured on Linux, but it was slow.
If you have the model in GPTQ format you have to convert it to GGML, but there might be some current issues with that: https://github.com/ggerganov/llama.cpp/issues/442

**Reply by dustyny** [reply]
Thanks @trollkotze I get it now..

**Comment by basujindal** [comment]
Hello, my apologies if it's not the right question/request, but can someone put a TLDR? 
I saw a lot of back and forth between whether the model needs 30GB or much less while maintaining similar tokens per second. AFAIK we need the entire model in RAM for fast inference on the CPU. Thanks!

**Reply by chadbrewbaker** [reply]
Most simple benchmark is an AWS CodeBuild script. ulimit the RAM and give latency decreases down to out of memory failure. Also do MPI localhost with two pinned cores with simulated network latency on the data transfer.

**Reply by SlyEcho** [reply]
It still needs the same amount of memory, except the OS may report it to the user differently when it's using memory-mapped files and not regular RAM.

**Reply by Green-Sky** [reply]
what I said here should still hold true overall. https://github.com/ggerganov/llama.cpp/discussions/638#discussioncomment-5482942

**Reply by basujindal** [reply]
Thank you for the clarifications.
Also, @SlyEcho  https://github.com/ggerganov/llama.cpp/discussions/638#discussioncomment-5492574 says that "My best guess is that my recent change caused "true" memory usage to decrease by 2x, because earlier, you had to make a full copy". Can someone explain that, is that really the case?

**Reply by abshkbh** [reply]
Haven't followed the entire discussion but I"ll try to explain mmap in this context if it helps in future discussions.

If a file is mmap-ed -

1. Virtual Page Addresses will be reserved for it i.e. think of it as book keeping cost for accessing the file. 

2. As the file is read at different offsets by your process, the relevant data from the file on disk is read into physical memory pages (page cache). Think of pages as 4KB of contiguous memory. 

3. As data is read into physical memory, your process' Virtual Addresses point to the physical pages in step 2. Now your process can access the data. 

4. The next time this data is read, we don't repeat the process we directly access the physical memory page. 

5. So you only need to load data from the file as you need it. Sometimes the OS will be smart and read ahead the next few pages as well to anticipate your request. 

6. You may have 8GB of RAM, what that means is that if your model needs more, your OS would need to keep evicting pages and swapping in data that's needed for the current request. 

7. If you have RAM larger than the model requires then you can keep it in RAM and pay the disk penalty only once.

8. Also, if you run the model once with decent RAM, chances are the model is in the page cache and subsequent runs won't pay the disk penalty to load the model.

9. You can also fork your process and your child will access the same memory address pointing to the model file. 


To conclude, mmap doesn't magically "reduce memory". It provides a memory interface for a file and a direct command to the OS to bring the file into its page cache / physical memory. 

Once the data is in page cache different tools may construe it as "memory not belonging to the process". But in reality, it does correspond to the data you need for your model to run.

Ordinary file reads via read() also cache file blocks in the OS (technically an OS can be stupid and not do this).

**Reply by Green-Sky** [reply]
adding to 
> 9. You can also fork your process and your child will access the same memory address pointing to the model file.

this also counts for multiple processes, assuming it is not modified by eg. lora .

**Reply by abshkbh** [reply]
Yes, the page cache is global and free to serve other processes as well
